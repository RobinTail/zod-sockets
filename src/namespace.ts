import { z } from "zod";
import { Emission, EmissionMap } from "./emission";
import { Hooks } from "./hooks";

export const rootNS = "/";
export type RootNS = typeof rootNS;

export const normalizeNS = (name: string): string => {
  const trimmed = name.trim();
  return trimmed.startsWith("/") ? trimmed : `/${trimmed}`;
};

interface Examples<T extends Emission> {
  payload?: z.output<T["schema"]>;
  ack?: T["ack"] extends z.AnyZodTuple ? z.input<T["ack"]> : never;
}

export interface Namespace<E extends EmissionMap, D extends z.SomeZodObject> {
  /**
   * @desc The events that the server can emit
   * @default {}
   * */
  emission: E;
  examples?: {
    [K in keyof E]?: Examples<E[K]> | Examples<E[K]>[];
  };
  /**
   * @desc Handlers for some events in different contexts
   * @default {}
   * */
  hooks: Hooks<E, D>;
  /**
   * @desc Schema of the client metadata in this namespace
   * @default z.object({})
   * */
  metadata: D;
}

export type Namespaces = Record<
  string,
  Namespace<EmissionMap, z.SomeZodObject>
>;
