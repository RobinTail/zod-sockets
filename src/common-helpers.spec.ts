import { describe, expect, test } from "vitest";
import { z } from "zod";
import {
  hasCoercion,
  lcFirst,
  makeCleanId,
  makeCleanIdWithFallback,
} from "./common-helpers";
import { rootNS } from "./namespace";

describe("Common helpers", () => {
  describe("hasCoercion()", () => {
    test.each([
      { schema: z.string(), coercion: false },
      { schema: z.coerce.string(), coercion: true },
      { schema: z.boolean({ coerce: true }), coercion: true },
      { schema: z.custom(), coercion: false },
    ])(
      "should check the presence and value of coerce prop %#",
      ({ schema, coercion }) => {
        expect(hasCoercion(schema)).toBe(coercion);
      },
    );
  });

  describe("makeCleanId()", () => {
    test.each([
      ["get"],
      ["post", "/", "something"],
      ["delete", "/user", "permanently"],
      ["patch", "/user/affiliated/account"],
      ["put", "/assets/into/:storageIdentifier"],
      ["get", "/flightDetails/:from-:to/:seatID"],
      ["get", "/companys/:companyId/users/:userId"],
    ])(
      "should generate valid identifier from the supplied strings %#",
      (...args) => {
        expect(makeCleanId(...args)).toMatchSnapshot();
      },
    );
  });

  describe("makeCleanIdWithFallback", () => {
    test.each([rootNS, "/", " / ", "."])(
      "should use fallback if it's too clean",
      (subject) => {
        expect(makeCleanIdWithFallback(subject, "Root")).toBe("Root");
      },
    );
    test("should make a clean id out of it", () => {
      expect(makeCleanIdWithFallback("/public/chat", "no")).toBe("PublicChat");
    });
  });

  describe("lcFirst()", () => {
    test("should make the first letter lower case", () => {
      expect(lcFirst("HereIsSomeText")).toBe("hereIsSomeText");
    });
  });
});
