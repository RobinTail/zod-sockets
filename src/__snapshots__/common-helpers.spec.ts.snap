// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Common helpers > getMessageFromError() > should compile a string from ZodError 1`] = `"user/id: expected number, got string; user/name: expected string, got number"`;

exports[`Common helpers > getMessageFromError() > should handle [ +0 ] path in ZodIssue 1`] = `"0: Custom error"`;

exports[`Common helpers > getMessageFromError() > should handle [ 1, 'some' ] path in ZodIssue 1`] = `"1/some: Custom error"`;

exports[`Common helpers > getMessageFromError() > should handle [] path in ZodIssue 1`] = `"Custom error"`;

exports[`Common helpers > getMessageFromError() > should pass message from other error types 0 1`] = `"something went wrong"`;

exports[`Common helpers > getMessageFromError() > should pass message from other error types 1 1`] = `"something went wrong"`;

exports[`Common helpers > makeCleanId() > should generate valid identifier from the supplied strings 0 1`] = `"Get"`;

exports[`Common helpers > makeCleanId() > should generate valid identifier from the supplied strings 1 1`] = `"PostSomething"`;

exports[`Common helpers > makeCleanId() > should generate valid identifier from the supplied strings 2 1`] = `"DeleteUserPermanently"`;

exports[`Common helpers > makeCleanId() > should generate valid identifier from the supplied strings 3 1`] = `"PatchUserAffiliatedAccount"`;

exports[`Common helpers > makeCleanId() > should generate valid identifier from the supplied strings 4 1`] = `"PutAssetsIntoStorageIdentifier"`;

exports[`Common helpers > makeCleanId() > should generate valid identifier from the supplied strings 5 1`] = `"GetFlightDetailsFromToSeatId"`;

exports[`Common helpers > makeCleanId() > should generate valid identifier from the supplied strings 6 1`] = `"GetCompanysCompanyIdUsersUserId"`;
